Simplify CreateEffectWarhead code

reaperrr opened the pull request:
My about half-dozenth attempt, and hopefully the first successful and therefore last one.

Simplified and streamlined code based on past feedback and suggestions.

Note: The new methods will move to `Warhead` later, once they're used by more than one warhead (I've come up with a use-case for that, but will leave that for a follow-up).

~Depends on #18066 and #18068.~

Closes #16308.

reaperrr committed:
Simplify CreateEffectWarhead code

Simplified and streamlined code,
based on past feedback and suggestions.

Note: The new methods will move to
Warhead later, once they're used by more
than one warhead.

pchote commented on OpenRA.Mods.Common/Warheads/Warhead.cs:
```diff
@@ -16,30 +16,28 @@
 		public readonly BitSet<TargetableType> InvalidTargets;
 
+		[Desc("What types of terrain targets are affected. Leave empty to affect all terrain not explicitly listed under InvalidTerrainTargets.")]
+		public readonly BitSet<TargetableType> ValidTerrainTargets;
```
Are these actually needed anymore? #18068 seems to have done fine without them.

reaperrr replied:
Eh, looks like I made a rebase error or pushed the wrong branch. Yeah, I meant to drop those.

reaperrr commented:
I'll fix and rebase this once the dependency is merged.

reaperrr commented:
Updated and rebased.

As far as testing goes, all you need to look out for is if there are any warhead explosion regressions vs bleed (explosions not showing, or multiple explosions showing even though yaml suggests only one should show on that actor/terrain).
It's basically #18068 all over again, except this time regressions are less likely, and would probably affect several places in RA and therefore be easier to spot.

abcdefg30 commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -135,26 +108,16 @@ public override void DoImpact(Target target, WarheadArgs args)
-			}
+			var dat = world.Map.DistanceAboveTerrain(pos);
+			var tileInfo = world.Map.GetTerrainInfo(cell);
+			return IsValidTarget(dat > AirThreshold ? TargetTypeAir : tileInfo.TargetTypes);
```
Fwiw, the tile info lookup is relatively expensive, so it might make sense to inline the variable here.
`return IsValidTarget(dat > AirThreshold ? TargetTypeAir : world.Map.GetTerrainInfo(cell).TargetTypes);`

pchote commented:
Any progress here @reaperrr?

I'd really like to get this in before the next prep branch so players can test the complete set of weapon changes as a single unit.

reaperrr commented:
I was thinking about using #16308 to claim redeployed bounties, which is why I didn't update it earlier.
They instead **un**deployed my bounties though, so that is no longer necessary.
Updated.

abcdefg30 commented:
This will need a rebase again eventually.

abcdefg30 commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -48,55 +48,26 @@ public class CreateEffectWarhead : Warhead
+			var potentialVictims = world.FindActorsOnCircle(pos, WDist.Zero);
+			foreach (var victim in potentialVictims)
 			{
-				if (!AffectsParent && victim == firedBy)
```
Is dropping this intentional? Seems like we don't use `AffectsParent` in this file at all any more.

reaperrr replied:
I honestly don't remember, but it probably was since `AffectsParent` defaulting to `false` has always been a horrible source of sublime (yaml) bugs for this warhead type, as `false` almost never makes sense for audiovisual effects.

I'll restore this, but make this warhead override the default instead.

reaperrr commented:
Rebased and updated.

abcdefg30 commented:
> Warheads/CreateEffectWarhead.cs(50,7): error SA1206: The 'public' modifier should appear before 'new' [/home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]

reaperrr commented:
Fixed.

Mailaender reviewed (dismissed):
I can't spot any regressions in-game.

reaperrr commented:
Rebased.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20200503/UpdateEffectWarheadTargeting.cs:
```diff
@@ -0,0 +1,39 @@
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			yield return "Note: due to larger internal refactoring of the CreateEffectWarhead code,\n" +
```
This would be better presented in the [SDK update notes](https://github.com/OpenRA/OpenRAModSDK/wiki/Update-notes:-Release-20200503-to-next-release), IMO. The updater instructions should really be limited to concrete instructions, and having this info here means that the modders who insist on not running the updater will never be able to see it!

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -46,57 +46,34 @@ public class CreateEffectWarhead : Warhead
 
-			foreach (var victim in victims)
+			// Check whether the impact position overlaps with an actor's hitshape
+			var potentialVictims = world.FindActorsOnCircle(pos, WDist.Zero);
```
Style nit: this is relatively short and only used once, so could move inline on the `foreach`

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -107,10 +84,12 @@ public override void DoImpact(Target target, WarheadArgs args)
+			var actorAtImpact = ImpactActors ? ActorTypeAtImpact(world, pos, firedBy) : ImpactActorType.None;
 
-			if ((!world.Map.Contains(targetTile)) || (!isValid))
+			// If there's either a) an invalid actor, or b) no actor and invalid terrain, we don't trigger the effect(s).
```
Style nit: This comment is a bit misleading IMO (an invalid actor is allowed provided there is also a valid actor), and the `else` in the `else if` is redundant.

How about writing this as:
```csharp
// Ignore the impact if there are only invalid actors within range
if (actorAtImpact == ImpactActorType.Invalid)
	return;

// Ignore the impact if there are no valid actors and no valid terrain
// (impacts are allowed on valid actors sitting on invalid terrain!)
if (actorAtImpact == ImpactActorType.None && !IsValidAgainstTerrain(world, pos))
	return;
```
which is exactly the same code, but IMO presented a bit more clearly?

pchote reviewed (commented):
I found one regression while testing:
* RA aircraft/helicopter husks no longer explode when crashing into the ground

Otherwise LGTM - the code makes sense and I didn't notice any other problems while testing. Just a couple of easy style nits:

reaperrr commented:
> RA aircraft/helicopter husks no longer explode when crashing into the ground

Ugh, that's _exactly_ what the `new` `AffectsParent` override was meant to prevent...
Guess I'll have to debug if the "new" makes it properly override the inherited default at all.

reaperrr commented:
> Guess I'll have to debug if the "new" makes it properly override the inherited default at all.

It doesn't. Anyone got an idea why
`public new readonly bool AffectsParent = true;`

doesn't seem to override the default `false` inherited from `Warhead`?

pchote commented:
```
UnitExplodePlane:
	Inherits: UnitExplode
	Warhead@2Eff: CreateEffect
		AffectsParent: true
		Explosions: large_napalm

UnitExplodeHeli:
	Inherits: UnitExplode
	Warhead@2Eff: CreateEffect
		AffectsParent: true
		Explosions: napalm
```
didn't help, so I think the problem might be something else?

reaperrr committed:
Fix airborne husk target types

Doesn't really make a difference, since it only matters
for effect warheads and those already could target both
air and ground actors, but GroundActor was still wrong.

reaperrr commented:
> didn't help, so I think the problem might be something else?

That's weird, in my tests it helped...

Anyway, updated. I'm 99.99% sure the problem was that `IsValidAgainst` by default checks `AffectsParent` as well, and since the effect warhead previously didn't override that method, it took the default from `Warhead`, where it defaults to `false`.

Since `CreateEffectWarhead` performs its own `AffectsParent` check before checking `IsValidAgainst`, I just dropped the check from the `IsValidAgainst` override.

Tested with RA Yak and Transport heli, seems to work as intended.

Edit: Forgot to adress the style nits, working on that now.

reaperrr commented:
Fixed style nits.

Salvaged text for the SDK update notes:

> Due to larger internal refactoring of the `CreateEffectWarhead` code, certain edge-case setups like explosions on submerged subs or aircraft husk explosions may no longer work as expected. You may need to update your `ValidTargets`/`InvalidTargets` to reconfigure such setups.
Generally speaking, for mods that use different explosions for different terrain and actor types, using separate `TargetTypes` for terrain and actors is now highly recommended. See the RA mods' updated actor `TargetTypes` and warhead `(In)ValidTargets` for examples.

pchote commented:
Another aircraft regression, sorry:
* RA aircraft/helicopter husks explode instead of splashing when hitting the water.

reaperrr commented:
Fixed by removing the `AffectsParent` default override.

Explanation:
In hindsight, the `AffectsParent` override was just a work-around for the real issue, which was that for the effect warhead, `AffectsParent: false` should have made the `IsValidAgainst` check _ignore_ the parent, not treat it as an 'invalid target'.
The `foreach` loop already did that correctly, but the `IsValidAgainst` check didn't (until the previous update).
For aircraft husks, `AffectsParent: true` would always result in either showing an explosion (if the parent was a valid target) or nothing (if it was aninvalid target), since crashing husks are obviously always at the impact position, so this needed to treat the parent as thin air via `AffectsParent: false` (but still needs to consider other actors, to show explosion instead of splash when the husk happens to hit a ship or naval building, for example).

Now `AffectsParent: false` works like `ImpactActors: false`, but limited to the parent actor.
We can still make the parent an 'invalid target' by setting `AffectsParent: true` and giving the actor a blacklisted target type, so this now allows us to cover all cases.

In the future, we could think about merging `ImpactActors` and `AffectsParent` into an enum (or flag list), but I don't want to risk any more delays or regressions for this particular PR (nor do I want to write an update rule now).

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -48,55 +48,44 @@ public class CreateEffectWarhead : Warhead
+
+		// We already check AffectsParent beforehand to avoid parent HitShape look-ups.
+		// Furthermore, the base method would read the 'false' default from Warhead and
+		// ignore the new default if we didn't override this method here.
```
Comment is still referring to the new default that no longer exists.

pchote commented:
LGTM otherwise, but we probably need another :+1: after the recent changes.

reaperrr commented:
Fixed comment.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e9ce63aa25d4add14fbeb54acad9ff878208161a...cf1d0cb3f8e5ec37e1966dbda3bdb716155d263a
